### PR TITLE
#12056 Update dis.findlinestarts for Python 3.13

### LINKED ISSUE #12056: Update for Python 3.13 builtins.TypeError: '<' not supported between instances of 'NoneType' and 'int'
I now see there is 3.13 alpha build available on GitHub https://github.com/actions/python-versions/releases/tag/3.13.0-alpha.2-6966007093

There is this run reports from Karolina for Fedora:

https://download.copr.fedorainfracloud.org/results/@python/python3.13/fedora-rawhide-x86_64/06770350-python-twisted/builder-live.log.gz

```
[FAIL]
Traceback (most recent call last):
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/test/test_assertions.py", line 1599, in test_callDeprecationWithWrongMessage
    exception = self.assertRaises(
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/_synctest.py", line 430, in assertRaises
    return context._handle(lambda: f(*args, **kwargs))
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/_synctest.py", line 320, in _handle
    with self as context:
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/_synctest.py", line 355, in __exit__
    self._testCase.fail(
twisted.trial.unittest.FailTest: builtins.TypeError raised instead of FailTest:
 Traceback (most recent call last):
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/python/util.py", line 958, in runWithWarningsSuppressed
    return f(*args, **kwargs)
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/test/test_assertions.py", line 1599, in test_callDeprecationWithWrongMessage
    exception = self.assertRaises(
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/_synctest.py", line 430, in assertRaises
    return context._handle(lambda: f(*args, **kwargs))
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/_synctest.py", line 320, in _handle
    with self as context:
--- <exception caught here> ---
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/_synctest.py", line 321, in _handle
    self._returnValue = obj()
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/_synctest.py", line 430, in <lambda>
    return context._handle(lambda: f(*args, **kwargs))
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/_synctest.py", line 1297, in callDeprecated
    warningsShown = self.flushWarnings([self.callDeprecated])
  File "/builddir/build/BUILD/twisted-twisted-23.10.0/src/twisted/trial/_synctest.py", line 1196, in flushWarnings
    if not (min(lineNumbers) <= aWarning.lineno <= max(lineNumbers)):
builtins.TypeError: '<' not supported between instances of 'NoneType' and 'int'


twisted.trial.test.test_assertions.CallDeprecatedTests.test_callDeprecationWithWrongMessage
```

Try to run the tests with py3.13 and see if we can fix this error.

We might not merge the py3.13 test run just yet.

### PR DESCRIPTION
**Pull request by adiroiban**
## Scope and purpose

Fixes #12056 

This updates the usage of dis.findlinestarts.

The Python 3.13 tests are cherry picked to be executed.
Later we can enable more.

As a drive-by, update GitHub Actions Python helper to the latest version.

### CODE REVIEW

**Comment by adiroiban** [comment]
This is ready for review, but before reviewing this, please take a look at https://github.com/twisted/twisted/pull/12058

Once we remove the cgi dependency, we can "unlock" many more tests on Python 3.13

needs-review

File: src/twisted/newsfragments/12056.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Update `int` and `None` comparisons as required by Python 3.13.
```
Comment (exarkun) [inline src/twisted/newsfragments/12056.feature]: Seems like a minor change.  As a user, what would this mean to me?  Just some internal implementation details.  "Support Python 3.13" will be a more interesting news fragment, when we get there. :)

Comment (adiroiban) [reply src/twisted/newsfragments/12056.feature]: True. Thanks. This info is wrong.

The change is in the way we use the result of `findlinesstarts` as mentioned in the title of this pr.

I converted this to a misc.

**Comment by befeleme** [comment]
I can confirm that using this patch on our attempted build for Fedora Linux makes all of the failures with  `builtins.TypeError: '<' not supported between instances of 'NoneType' and 'int'` go away - the tests pass again.

**Comment by eevelweezel** [comment]
I can confirm that this change works with my changes in #12061, allowing tests in twisted.internet to pass on Python 3.13 (excluding those that require cffi and other yet-unsupported dependencies).

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -134,6 +134,21 @@ jobs:
...
           - python-version: '3.11'
 
+          # Just Python 3.13 alpha with default settings.
+          # For now, minimum dependencies as cffi doesn't work on 3.13.
+          # No coverage also.
+          - python-version: '3.13.0-alpha.2'
+            tox-env: 'mindeps-nocov-posix'
+            job-name: 'mindeps-nocov-3.13'
+            skip-coverage: yes
+            # FIXME:https://github.com/twisted/twisted/issues/12060
```
Comment (eevelweezel) [inline .github/workflows/test.yaml:144]: I can confirm that this change works with my changes in https://github.com/twisted/twisted/issues/12061, allowing tests in twisted.internet to pass on Python 3.13 (excluding those that require cffi and other yet-unsupported dependencies).

**Review by eevelweezel — APPROVED** [review]

**Comment by adiroiban** [comment]
Thanks for your reviews.

@exarkun I have updated the release notes and automated tests so that we can have part of the curren test running on 3.13.

Do you have time to take another look at this?

With future PRs we can enable more tests.

I think is important to have this merged as the base for future py3 changes.

**Review by exarkun — APPROVED** [review]

**Comment by adiroiban** [comment]
Many thanks for your reviews. I have merged this.
Now, when we work in Python 3.13 PR we can enable the test for the parts that were fixed.

**Pull request merged**